quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,73 @@
+//! Pacing of packet transmissions.
+
+use std::{
+    convert::TryInto,
+    time::{Duration, Instant},
+};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate determined by
+/// the congestion window and estimated round-trip time.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    last_instant: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            last_instant: Instant::now(),
+        }
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns [`None`]. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// We panic if the passed-in instant is older than `self.last_instant`.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        mut now: Instant,
+    ) -> Option<Instant> {
+        if let Some(time_elapsed) = now.checked_duration_since(self.last_instant) {
+            self.last_instant = now;
+            let new_tokens = window.saturating_mul(5) / 4;
+            if smoothed_rtt.as_nanos() == 0 {
+                return None;
+            }
+            let elapsed = (time_elapsed.as_nanos() / smoothed_rtt.as_nanos())
+                .try_into()
+                .unwrap_or(u64::max_value());
+
+            let new_tokens = new_tokens.saturating_mul(elapsed);
+            self.tokens = self.tokens.saturating_add(new_tokens).min(self.capacity);
+        } else {
+            now = self.last_instant
+        }
+
+        if window > u32::max_value().into() {
+            // we disable pacing for extremely large windows
+            return None;
+        }
+
+        let window = window as u32;
+        if self.tokens > mtu.into() {
```

Comment:
I think turning this into an early return would be a bit nicer. Maybe move that directly after the assignment to `self.tokens`?